File: matlab/Handlers.m
classdef Handlers
    properties(Constant)
        EA = EA_Engine();
        engine = Handlers.EA.engine;
    end
    methods (Static)
        % static variables
        function timestamp = StaticTimestamp(newTimestamp)
            persistent currentTimestamp
            if nargin
                currentTimestamp = newTimestamp;
            end
            if isempty(currentTimestamp)
                currentTimestamp = 0;
            end
            timestamp = currentTimestamp;
        end
        function curves = StaticCurves(newCurves)
            persistent currentCurves
            if nargin
                currentCurves = newCurves;
            end
            if isempty(currentCurves)
                currentCurves = cell(1, 10);
            end
            curves = currentCurves;
        end
        function databuffer = StaticDataBuffer(newDataBuffer)
            persistent currentDataBuffer
            if nargin
                currentDataBuffer = newDataBuffer;
            end
            if isempty(currentDataBuffer)
                fs = 96e3;
                duration = 10;
                currentDataBuffer = Buffer(fs * duration);
            end
            databuffer = currentDataBuffer;
        end
        function channelInfo = StaticChannelInfo(newChannelInfo)
            persistent currentChannelInfo
            if nargin
                currentChannelInfo = newChannelInfo;
            end
            if isempty(currentChannelInfo)
                currentChannelInfo = cell(10,3);
            end
            channelInfo = currentChannelInfo;
        end
        function testResults = StaticTestResults(newTestResults)
            persistent currentTestResults
            if nargin
                currentTestResults = newTestResults;
            end
            if isempty(currentTestResults)
                currentTestResults = [];
            end
            testResults = currentTestResults;
        end

        % Static methods
        function initDataBuffer(buffer)
            Handlers.StaticDataBuffer(buffer);
        end

        function averageUpdatedHandler(source, args)
            if args.CurrentAverage < args.TotalAverages
                fprintf('Average: %d / %d\n', args.CurrentAverage, args.TotalAverages);
            else
                fprintf('Average: %d / %d\n', args.TotalAverages, args.TotalAverages);
                fprintf('\n');
            end
        end
        function calfeedbackHandler(source, args)
            fprintf('%s - %s\n', char(args.MessageType), char(args.Message));
        end
        function feedbackHandler(source, args)
            fprintf('%s - %s\n', char(args.MessageType), char(args.Message));
        end
        function timedataHandler(source, args)
            if ~isempty(args.TimeBlocks)
                data = args.TimeBlocks;

                n = 0;

                if data.Length ~= 0
                    for j = 1:data.GetLength(2)
                        % y = zeros(1, 4096);
                        y = cell(1, 4096); % Initialize a cell array to hold the data

                        for i = 1:data.GetLength(1)
                            y{i} = data.Get(i - 1, j - 1); % Adjusted indexing for MATLAB
                        end

                        % Convert cell array to numeric array before setting data
                        curves = Handlers.StaticCurves();
                        curves{n + 1}.YData = cell2mat(y);
                        curves{n + 1}.XData = timestamp;
                        Handlers.StaticCurves(curves);
                        n = n + 1;
                    end
                end
            end
        end
        function timeUpdatedHandler(source, args)
            if args.CurrentTime < args.Duration
                fprintf('%s: %.2f\n', char(args.CurrentMessage), args.CurrentTime);
                timestamp = args.CurrentTime;
                Handlers.StaticTimestamp(timestamp);
            else
                fprintf('\n');
            end
        end
        function fftprocessingUpdated(source, args)
            persistent plotTick
            if isempty(plotTick)
                plotTick = 0;
            end
            if ~isempty(args.AutoSpectrum)
                curves = cell(1,args.AutoSpectrumInfo.Length);
                autoSpectrumInfo = cell(args.AutoSpectrumInfo.Length,1);
                for i = 1:args.AutoSpectrumInfo.Length
                    autoSpectrumInfo{i} = char(args.AutoSpectrumInfo.GetValue(i-1));
                end
                channelInfo = Handlers.StaticChannelInfo();
                for i = 1:length(autoSpectrumInfo)
                    splitInfo = strsplit(autoSpectrumInfo{i}, '|');
                    for j = 1:length(splitInfo)
                        channelInfo{i, j} = splitInfo{j};
                    end
                end
                Handlers.StaticChannelInfo(channelInfo);
                freq = args.FrequencyAxis;
                data = args.AutoSpectrum;
                if ~isempty(freq)
                    for j = 1:length(curves)
                        x = zeros(freq.Length, 1);
                        y = zeros(freq.Length, 1);
                        for i = 1:freq.Length
                            x(i) = freq(i);
                            y(i) = data(i, j);
                        end
                        curves{j} = {x, y};

                    end
                    if plotTick <= 0  
                        plotTick = 5;
                        Handlers.frequencyPlots();
                    end
                    plotTick = plotTick - 1;
                end
            end
            Handlers.StaticCurves(curves);    
        end
        function stepProcessingUpdated(source, args)
            persistent plotTick
            if isempty(plotTick)
                plotTick = 0;
            end

            if ~isempty(args.SpectrumAmplitude)
                curves = cell(1, length(args.SpectrumInfo));
                spectrumInfo = cell(length(args.SpectrumInfo), 1);
                for i = 1:length(args.SpectrumInfo)
                    spectrumInfo{i} = char(args.SpectrumInfo.GetValue(i-1));
                end
                for i = 1:length(spectrumInfo)
                    splitInfo = strsplit(spectrumInfo{i}, '|');
                    for j = 1:length(splitInfo)
                        channelInfo = Handlers.StaticChannelInfo();
                        channelInfo{i, j} = splitInfo{j};
                    end
                end
                Handlers.StaticChannelInfo(channelInfo);
                freq = args.FrequencyAxis;
                data = args.SpectrumAmplitude;
                if ~isempty(freq)
                    for j = 1:length(curves)
                        x = zeros(freq.Length, 1);
                        y = zeros(freq.Length, 1);
                        for i = 1:freq.Length
                            x = freq;
                            y(i) = data(i, j);
                        end
                        curves{j} = {x, y};
                    end
                    if plotTick <= 0  
                        plotTick = 5;
                        Handlers.frequencyPlots();
                    end
                    plotTick = plotTick - 1;
                end
            end
            Handlers.StaticCurves(curves);
        end
        function time_data_callback(source, args)
            if ~isempty(args.TimeBlocks)
                data = args.TimeBlocks;
                if data.GetLength(0) ~= 0
                    for j = 1:data.GetLength(1)
                        y = zeros(1, 4096);
                        for i = 1:data.GetLength(0)
                            y(i) = data(i, j);
                        end
                        DataBuffer = Handlers.StaticDataBuffer();
                        DataBuffer = DataBuffer.append(y);
                        Handlers.StaticDataBuffer(DataBuffer);
                    end
                end
            end
        end
        function frequencyPlots()
            fig = findobj('Type', 'Figure', 'Name', 'FFT Plot');
            if isempty(fig)
                figure('Name', 'FFT Plot');
            else
                figure(fig);
            end
            curves = Handlers.StaticCurves();
            channelInfo = Handlers.StaticChannelInfo();
            for i = 1:length(curves)
                subplot(length(curves), 1, i);
                try
                    semilogx(curves{i}{1}, curves{i}{2});
                catch
                    return;
                end
                grid on;
                xlabel('Frequency');
                ylabel('Amplitude (dB)');
                title(sprintf('%s: %s', channelInfo{i, 1}, channelInfo{i, 3}));
            end
            drawnow;
        end
        function stimulusCreatedHandler(source, args)
            import EA_Engine.*
            EA = EA_Engine();
            engine = EA.engine;
            outputChannelName = engine.GetSelectedWDMOutputDevice().Name.Replace(" Left", "").Replace(" Right", "");
            deviceId = NaN;
            devices = audiodevinfo().output;
            for i = 1:length(devices)
                if strfind(string(devices(i).Name), string(outputChannelName))
                    deviceId = devices(i).ID;
                    break
                end
            end
            wavPath = string(args.FullPath);
            [audioData, sampleRate] = audioread(wavPath);
            persistent player;
            player = audioplayer(audioData, sampleRate, 24, deviceId);
            player.play();
        end

        function testResultsReadyHandler(source, args)
            if ~isempty(args.AutoSpectrum)
                curves = cell(1,args.AutoSpectrumInfo.Length);
                autoSpectrumInfo = cell(args.AutoSpectrumInfo.Length,1);
                for i = 1:args.AutoSpectrumInfo.Length
                    autoSpectrumInfo{i} = char(args.AutoSpectrumInfo.GetValue(i-1));
                end
                channelInfo = Handlers.StaticChannelInfo();
                for i = 1:length(autoSpectrumInfo)
                    splitInfo = strsplit(autoSpectrumInfo{i}, '|');
                    for j = 1:length(splitInfo)
                        channelInfo{i, j} = splitInfo{j};
                    end
                end
                Handlers.StaticChannelInfo(channelInfo);
                freq = args.FrequencyAxis;
                data = args.AutoSpectrum;
                if ~isempty(freq)
                    for j = 1:length(curves)
                        x = zeros(freq.Length, 1);
                        y = zeros(freq.Length, 1);
                        for i = 1:freq.Length
                            x(i) = freq(i);
                            y(i) = data(i, j);
                        end
                        curves{j} = {x, y};

                    end
                    Handlers.frequencyPlots();
                end
            end
            Handlers.StaticCurves(curves);    
            testResults = args;
            Handlers.StaticTestResults(testResults);
        end

        function saveResultsToVarHandler(source, args)
            testResults = args;
            Handlers.StaticTestResults(testResults);
        end
    end
end

File: csharp/csharp/Handlers.cs
using EA_Engine;
using NAudio.Wave;

namespace EAProject
{
	public static class Handlers
	{
        public static double timestamp = 0;
        public static (double[], double[])[] dataPoints;
        public static bool frequencyPlotReady = false;
        public static string[,] channelInfo = new string[10,3];
        public static TestResultsAvailableArgs? testResults;

		private static readonly Lazy<Engine> LazyEngine = new Lazy<Engine>(() =>  HelperEngine.GetEngine());
        private static Engine engine = LazyEngine.Value;

        public static void TimeUpdatedHandler(object sender, TimeUpdatedArgs e)
        {
            if (e.CurrentTime < e.Duration)
            {
                Console.Write($"\r{e.CurrentMessage}: {e.CurrentTime:F2}");
                timestamp = e.CurrentTime;
            }
            else
            {
                Console.WriteLine();
            }
        }
        public static void FeedbackHandler(object sender, FeedbackdArgs e)
        {
            Console.WriteLine($"{e.MessageType} - {e.Message}");
        }
        public static void AverageUpdatedHandler(object sender, AverageUpdatedArgs e)
        {
            if (e.CurrentAverage < e.TotalAverages)
            {
                Console.Write($"\rAverage: {e.CurrentAverage} / {e.TotalAverages}");
            }
            else
            {
                Console.Write($"\rAverage: {e.TotalAverages} / {e.TotalAverages}");
                Console.WriteLine();
            }
        }
        public static void TimeDataCallback(object sender, TimeDataRecordedArgs args)
        {
            if (args.TimeBlocks != null)
            {
                var data = args.TimeBlocks;
                if (data.GetLength(0) != 0)
                {
					double[] y = new double[4096];
					for (int i = 0; i < data.GetLength(0); i++)
					{
						y[i] = data[i, 0];
					}
					Buffer.Append(y);
                }
            }
        }
        public static void FftProcessingUpdated(object sender, MultiSIMOFFTProcessingResultsUpdatedArgs args)
        {
            if (args.AutoSpectrum != null)
            {
				dataPoints = new (double[], double[])[args.AutoSpectrumInfo.Length];

				for (int i = 0; i < args.AutoSpectrumInfo.Length; i++)
				{
                    string[] splitInfo  = args.AutoSpectrumInfo[i].Split('|');
                    for (int j = 0; j < splitInfo.Length; j++)
                    {
                        channelInfo[i, j] = splitInfo[j];
                    }
				}
                double[] freq = args.FrequencyAxis;
                var data = args.AutoSpectrum;
                if (freq.Length != 0)
                {
                    for (int i = 0; i < data.GetLength(1); i++)
                    {
                        var x = new double[freq.Length];
                        var y = new double[freq.Length];
                        for (int j = 0; j < freq.Length; j++)
                        {
                            x[j] = freq[j];
                            var test = data[j, i];
                            y[j] = data[j, i];
                        }
                        dataPoints[i] = (x, y);
                    }
                }
            }
            frequencyPlotReady = true;
        }
        public static void StepProcessingUpdated(object sender, HarmonicEstimatorProcessingResultsUpdatedArgs args)
        {

            if (args.SpectrumAmplitude != null)
            {
				dataPoints = new (double[], double[])[args.SpectrumInfo.Length];

				for (int i = 0; i < args.SpectrumInfo.Length; i++)
				{
                    string[] splitInfo  = args.SpectrumInfo[i].Split('|');
                    for (int j = 0; j < splitInfo.Length; j++)
                    {
                        channelInfo[i, j] = splitInfo[j];
                    }
				}
                double[] freq = args.FrequencyAxis;
                var data = args.SpectrumAmplitude;
                if (freq.Length != 0)
                {
                    for (int i = 0; i < data.GetLength(1); i++)
                    {
                        var x = new double[freq.Length];
                        var y = new double[freq.Length];
                        for (int j = 0; j < freq.Length; j++)
                        {
                            x[j] = freq[j];
                            var test = data[j, i];
                            y[j] = data[j, i];
                        }
                        dataPoints[i] = (x, y);
                    }
                }
            }
            frequencyPlotReady = true;
        }
        
        public static void StimulusCreatedHandler(object sender, StimulusCreatedArgs args)
        {
            var outputChannelName = engine.GetSelectedWDMOutputDevice().Name.Replace(" Left", "").Replace(" Right", "");
            int? deviceNumber = null;
            
            for (int i = -1; i < WaveOut.DeviceCount; i++)
            {
                var caps = WaveOut.GetCapabilities(i);
                if (outputChannelName.StartsWith(caps.ProductName))
                {
                    deviceNumber = i;
                    break;
                }
            }

            var wavPath = args.FullPath;
            WaveFileReader wav = new WaveFileReader(wavPath);
            if (deviceNumber == null)
            {
                System.Console.WriteLine("ERROR: Could not find device to play stimulus");
                return;

			}
            var output = new WaveOutEvent{ DeviceNumber = deviceNumber.Value };
            output.Init(wav);
            output.Play();
        }

        public static void TestResultsReadyHandler(object sender, TestResultsAvailableArgs args)
        {
            testResults = null;
            if (args.AutoSpectrum != null)
            {
				dataPoints = new (double[], double[])[args.AutoSpectrumInfo.Length];

				for (int i = 0; i < args.AutoSpectrumInfo.Length; i++)
				{
                    string[] splitInfo  = args.AutoSpectrumInfo[i].Split('|');
                    for (int j = 0; j < splitInfo.Length; j++)
                    {
                        channelInfo[i, j] = splitInfo[j];
                    }
				}
                double[] freq = args.FrequencyAxis;
                var data = args.AutoSpectrum;
                if (freq.Length != 0)
                {
                    for (int i = 0; i < data.GetLength(1); i++)
                    {
                        var x = new double[freq.Length];
                        var y = new double[freq.Length];
                        for (int j = 0; j < freq.Length; j++)
                        {
                            x[j] = freq[j];
                            var test = data[j, i];
                            y[j] = data[j, i];
                        }
                        dataPoints[i] = (x, y);
                    }
                }
				frequencyPlotReady = true;
            }
			testResults = args;
        }
        public static void SaveResultsToVarHandler(object sender, TestResultsAvailableArgs args)
        {
            testResults = args;
        }
	}
}
